install/kubernetes/helm/istio/files/injection-template.yaml

```diff
@@ -24,6 +24,10 @@ initContainers:
   - "{{ annotation .ObjectMeta `traffic.sidecar.istio.io/includeInboundPorts` (includeInboundPorts .Spec.Containers) }}"
   - "-d"
   - "{{ excludeInboundPort (annotation .ObjectMeta `status.sidecar.istio.io/port` .Values.global.proxy.statusPort) (annotation .ObjectMeta `traffic.sidecar.istio.io/excludeInboundPorts` .Values.global.proxy.excludeInboundPorts) }}"
+  {{ if or (isset .ObjectMeta.Annotations `traffic.sidecar.istio.io/excludeOutboundPorts`) (ne .Values.global.proxy.excludeOutboundPorts "") -}}
```

Comment:
IMO, you should default excludeOutboundPorts in the template itself as well.   This seems to work:
```
{{ if or (isset .ObjectMeta.Annotations `traffic.sidecar.istio.io/excludeOutboundPorts`) (ne (valueOrDefault .Values.global.proxy.excludeOutboundPorts "") "") -}}
```

Background on this comment...  In another context, I have some general questions on how careful we need to be adding use of new vars to the injection-template.yaml due to potential users' use of manual injection with old values.yaml files and this template directly--e.g. the istio/installer demo test does this so it doesn't use the configmap with the helm rendered values loaded:

```
istioctl kube-inject -f test/simple/servicesToBeInjected.yaml \
        -n demo \
        --meshConfigFile test/demo/mesh.yaml \
        --valuesFile test/simple/values.yaml \
        --injectConfigFile istio-control/istio-autoinject/files/injection-template.yaml
```

Reply:
I don't understand what you mean by this? nothing in this file is rendered by helm

Reply:
I mean that variables used in this file that are not from annotations are added into the configmap via helm and the auto-sidecar injector loads them prior to rendering this template.  In the `istioctl kube-inject` case you either point it at the configmap or point it at a values file and this template like I mention istio/installer's test does.  So, adding a new variable to this template means it is undefined in existing values.yaml files and the kube-inject rendering will fail unless you use `valueOrDefault` when you reference it.

Reply:
I am pretty sure `values.global.proxy.somethingthatdoesntexist` === `""`, because of Go's default nil values. If you `.values.foo.bar` it would be an issue if foo isn't defined, but the last field will default to the nil value. I don't think this is a concern

Reply:
my test failures say otherwise :-)

Reply:
Wow I am surprised... I just tested this as well and seems you are right. I could have sworn we depend on this fact all over the place. Good catch

Reply:
yeah I was also hoping go template would save me when I was doing an installer change to add a var to this template.  So the pain was fresh :-)